aignostic/aignostic/router/api.py

```diff
@@ -1,23 +1,113 @@
-from pydantic import BaseModel
-from fastapi import APIRouter
+from pydantic import BaseModel, HttpUrl
+from fastapi import APIRouter, HTTPException
+import requests
+import aignostic.metrics.metrics as metrics_lib
+
 
 api = APIRouter()
 
 
+class DatasetRequest(BaseModel):
+    datasetURL: HttpUrl
+    modelURL: HttpUrl
+    metrics: list[str]
+
+
+@api.post("/evaluate")
+async def generate_metrics_from_info(request: DatasetRequest):
+    """
+    Frontend will post URLs, metrics, etc., as JSON to this endpoint.
+    This function validates, processes, and forwards the data to the controller.
+    """
+    # Extract data from the validated request
+    datasetURL = request.datasetURL
+    modelURL = request.modelURL
+    metrics = request.metrics
+
+    results = await process_data(datasetURL, modelURL, metrics)
+
+    return {"message": "Data successfully received", "results": results}
+
+
+# Sanity check endpoint
+# This for checking time of last deployment
 @api.get("/")
-def hello():
-    return {"message": "Hello Worlds!"}
+def info():
+    return {"message": "Pushed at 21/01/2025 07:32"}
+
+
+async def process_data(datasetURL: HttpUrl, modelURL: HttpUrl, metrics: list[str]):
+    """
+    Controller function. Takes data from the frontend, received at the endpoint and then:
+    - Passes to data endpoint and fetch data
+    - Process the data in preparation for passing to the model
+    - Pass to the model, and get the predicitons
+
+    Params:
+    - datasetURL : API URL of the dataset
+    - modelURL : API URL of the dataset
+    - metrics: list of metrics that should be applied
+    """
+
+    # fetch data from datasetURL
+    data: dict = await fetch_data(datasetURL)
+
+    # strip the label from the datapoint
+    try:
+        rows = data["rows"]
+        feature, true_label = [rows[0][:-1]], [rows[0][-1]]
+        prediction = await query_model(modelURL, {"column_names": data["column_names"][:-1],  "rows": feature})
+        predicted_labels = [item for sublist in prediction["rows"] for item in sublist]
+        metrics_results = metrics_lib.calculate_metrics(true_label, predicted_labels, metrics)
+    except Exception as e:
```

Comment:
This again catches keyboard interrupts but since it's a server anyway should be fine?